# Bài tập 2

Họ và tên: Bùi Chí Dũng - MSSV: 1712364

## Câu 1 (1 điểm)

In [1]:
import numpy as np

In [2]:
N = 10 
mu = 0.5 
n_repeats = 100000 
n_coins = 1000

In [3]:
nus = np.random.binomial(N,mu,(n_repeats, n_coins)) / N

In [4]:
np.mean(np.min(nus, axis = 1))

0.037738

Kết quả em chạy ra được là vmin = 0.037359. Vậy đáp án gần nhất là [b] 0.01

## Câu 2 (1 điểm)

In [5]:
rows = np.arange(nus.shape[0])
cols = np.random.choice(np.arange(nus.shape[1]), size = nus.shape[0])
nus[rows,cols]

array([0.6, 0.3, 0.7, ..., 0.2, 0.5, 0.7])

In [6]:
V1 = np.mean(nus[:, 0])
Vrand = np.mean(nus[rows,cols])
Vmin = np.mean(np.min(nus, axis=1))
print("V1 = ", V1)
print("Vrand = ", Vrand)
print("Vmin = ", Vmin)

V1 =  0.5000530000000001
Vrand =  0.5006879999999999
Vmin =  0.037738


Qua kết quả trên, ta thấy V1 và Vrand là gần bằng với mu=0.5 (vì fair coins nên các mu của c1,crand,cmin đều bằng nhau và bằng mu = 0.5). Chính vì vậy mà nó thỏa Hoeffding Inequality. Vậy, đáp án sẽ là [d] c1 and crand.

## Câu 3 (1 điểm)

Ta có hypothesis function h(x), là hàm mà chúng ta train trên target function y = f(x), nhưng nó phát sinh lỗi với xác suất là μ. Sau khi tìm được h(x) chúng ta sẽ áp dụng đối với tập dữ liệu (có noise, dữ liệu thực tế).
Xác suất h(x) sinh lỗi trong trường hợp dữ liệu không có nhiễu là μ, và xác suất ngược lại là 1 - μ.
Đối với xác suất gây ra lỗi do nhiễu là 1 - λ với λ là xác suất dữ liệu có nhiễu nhưng không sinh lỗi. Vì đây là hàm nhị phân nên xác suất sinh ra lỗi khi áp dụng h(x) vào f(x) có nhiễu chính là: 
- Xác suất phát sinh lỗi do nhiễu (1 - λ), và không phát sinh lỗi (1-μ)
hoặc
- Xác suất không sinh lỗi do nhiễu (λ), và có phát sinh lỗi (μ)
Vì vậy xác suất phát sinh lỗi khi h xấp xỉ y là:
P(error) = P(noise error) * P(no mu error) + P(no noise error) * P(mu error)
= (1 - λ) * (1-μ) + λ * μ

Ta được xác suất:
[e] (1 − λ) ∗ (1 − µ) + λ ∗ µ

## Câu 4 (1 điểm)

Giá trị của λ khi mà h độc lập với µ trong công thức μ(2λ−1)+1−λ khi μ bị triệt tiêu, cũng là khi 2λ−1 = 0. Dễ thấy 2λ−1 = 0 khi λ = 0.5.
Chọn [b] 0.5

## Câu 5 (1 điểm)

In [7]:
import numpy as np
d = 2 # Input vector dimentional (Trong đề là 2 chiều)
def generate_Separable_Data(N):
    #X: mảng Numpy, kích thước N x(d+1), N là số lượng mẫu huấn luyện, 
    #d là số chiều của vector input (chưa tính x0 = 1).
    X = 2*np.random.rand(N, d+1)-1
    X[:,0] = 1.0
    xa = 2*np.random.rand() - 1
    ya = 2*np.random.rand() - 1
    xb = 2*np.random.rand() - 1
    yb = 2*np.random.rand() - 1
    xa, xb = min(xa, xb), max(xa, xb)
    a = (yb-ya)/(xb-xa)
    ys = 2*(X[:,2] > ya + (X[:,1] - xa)*a)-1
    return X, ys

class LRB_Classifier:
    def __init__(self):
        pass
    
    def fit(self, X, ys):
        X_dagger = np.dot(np.linalg.pinv(np.dot(X.T, X)), X.T)
        self.w = np.dot(X_dagger,ys)
    # Lấy dấu của mảng dự đoán np.dot(X,self.w)
    def prediction(self, X):
        return np.sign(np.dot(X,self.w))

In [8]:
E_ins = []

for i in range(1000):
    # Tạo dữ liệu và true target function
    X, ys = generate_Separable_Data(100)

    # Tạo bộ phân lớp dựa trên hồi quy tuyến tính bình phương nhỏ nhất
    ls_classifier = LRB_Classifier()
    ls_classifier.fit(X, ys)

    # Tìm Error_in_sample
    y_predict = ls_classifier.prediction(X)
    E_in = np.mean(ys != y_predict)
    E_ins.append(E_in)
    
print("The average of Ein = ", np.mean(E_ins))

The average of Ein =  0.039810000000000005


Ta thấy, giá trị trung bình của E_in xung quanh 0.04, vậy đáp án gần nhất là [c] 0.01

## Câu 6 (1 điểm)

In [9]:
E_outs = []

for i in range(1000):
    #Tạo dữ liệu và true target function
    X, y = generate_Separable_Data(1000) #1000 fresh points
    Xs, ys = X[0:100,:], y[0:100] #Xs,ys (s = sample)
    Xr, yr = X[100:,:], y[100:] #Xr,yr (r = remaining)

    #Tạo bộ phân lớp dựa trên hồi quy tuyến tính bình phương nhỏ nhất
    ls_classifier = LRB_Classifier()
    ls_classifier.fit(Xs, ys)

    # Tìm Error_in_sample
    y_predict = ls_classifier.prediction(Xr)
    E_out = np.mean(yr != y_predict )
    #Cập nhật Eouts
    E_outs.append(E_out)
    
print("The average of Ein = ", np.mean(E_outs))

The average of Ein =  0.04947777777777777


Ta thấy, giá trị trung bình của E_in xung quanh 0.04 ~ 0.05, vậy đáp án gần nhất là [c] 0.01

## Câu 7 (1 điểm)

In [10]:
class PB_Classifier:
    def __init__(self):
        self.w = np.zeros((X.shape[1],1))
    
    def fit(self, X, y, init_w = None):
        self.w = np.zeros((X.shape[1],1))
        if not init_w is None:
            self.w = init_w
        
        self.n_iterations = 0
        while True:
            y_predict = self.prediction(X)
            mis = np.nonzero(y != y_predict)[0]
            if len(mis) == 0:
                return
            # choose a point randomly from the set of misclassified(mis) points at each iteration
            i = np.random.choice(mis) 
            self.w = self.w + y[i]*X[i,:]
            self.n_iterations += 1
            if self.n_iterations > 1000:
                print("PLA is impossible!")
                return
            
    def prediction(self, X):
        return np.sign(np.dot(X,self.w))
    
iters = []
for i in range(1000):
    #Tạo dữ liệu và true target function
    X, y = generate_Separable_Data(10)

    #Tạo bộ phân lớp dựa trên hồi quy tuyến tính bình phương nhỏ nhất
    ls_classifier = LRB_Classifier()
    ls_classifier.fit(X, y)
    
    #Tạo bộ phân lớp PLA
    pla = PB_Classifier()
    pla.fit(X, y, init_w = ls_classifier.w)
    iters.append(pla.n_iterations)
    
print("The average number of interations in over 1000 runs that PLA takes to converge = ", np.mean(iters))

The average number of interations in over 1000 runs that PLA takes to converge =  4.141


Ta thấy, giá trị trung bình các iterations trong 1000 lần chạy nằm ở mức gần bằng 3 ~ 5. Vì vậy đáp án gần nhất là [a] 1

## Câu 8 (1 điểm)

In [11]:
def f(x):
    return np.sign(x[:,1]**2 + x[:,2]**2 - 0.6)

def generate_Dataset(N):
    X = np.random.uniform(-1, 1, (N,3))
    X[:,0] = 1.0
    y = f(X)
    
    # Chọn ngẫu nhiên 10% các giá trị output đúng của y
    y_output_ten_percents = np.random.choice(np.arange(0,N),N//10)
    y[y_output_ten_percents] *= -1 #Đổi dấu y
    
    return X, y

In [12]:
E_ins = []
for i in range(1000):
    X, y = generate_Dataset(100)
    lsr = LRB_Classifier()
    lsr.fit(X, y)
    y_predict = lsr.prediction(X)
    E_ins.append(np.mean(y != y_predict))
    
print("The sample error E_in = : ", np.mean(E_ins))

The sample error E_in = :  0.43831


Ta thấy, tỉ lệ lỗi gần bằng 0.5, vậy chọn đáp án [d] 0.5

## Câu 9 (1 điểm)

In [13]:
def to_Z(X):
    Z = np.zeros((X.shape[0], 6))
    Z[:,0] = 1.0
    Z[:,1] = X[:,1]
    Z[:,2] = X[:,2]
    Z[:,3] = X[:,1]*X[:,2]
    Z[:,4] = X[:,1]**2
    Z[:,5] = X[:,2]**2
    return Z

X, ys = generate_Dataset(1000)
Z = to_Z(X)
lsr = LRB_Classifier()
lsr.fit(Z, ys)
w_tilde = lsr.w
print(w_tilde)

[-0.98712139  0.04115769  0.00916052  0.04996928  1.57844384  1.60217457]


Ta thấy các giá trị gần bằng -1, 0.05, 0.08, 0.13, 1.5, 1.5. Vậy ta chọn đáp án [a] g(x1, x2) = sign(−1 − 0.05x1 + 0.08x2 + 0.13x1x2 + 1.5x1*x1 + 1.5x2*x2

## Câu 10 (1 điểm)

In [14]:
E_outs = []

for i in range(1000):
    Xs, ys = generate_Dataset(2000)

    X_train, y_train = Xs[:1000,:], ys[0:1000]
    X_test, y_test = Xs[1000:,:], ys[1000:]

    Z_train = to_Z(X_train)
    Z_test = to_Z(X_test)

    lsr = LRB_Classifier()
    lsr.fit(Z_train, y_train)

    y_predict = lsr.prediction(Z_test)
    E_out = np.mean(y_test != y_predict)
    E_outs.append(E_out)
    
print("The sample error E_out = = :", np.mean(E_outs))

The sample error E_out = = : 0.122166


Tỉ lệ lỗi E_out gần bằng 0.12. Vậy, ta chọn đáp án gần nhất là [b] 0.1